agent/agentcontainers/watcher/watcher.go

```diff
@@ -0,0 +1,89 @@
+// Package watcher provides file system watching capabilities for the
+// agent. It defines an interface for monitoring file changes and
+// implementations that can be used to detect when configuration files
+// are modified. This is primarily used to track changes to devcontainer
+// configuration files and notify users when containers need to be
+// recreated to apply the new configuration.
+package watcher
+
+import (
+	"context"
+
+	"github.com/fsnotify/fsnotify"
+	"golang.org/x/xerrors"
+)
+
+// Watcher defines an interface for monitoring file system changes.
+// Implementations track file modifications and provide an event stream
+// that clients can consume to react to changes.
+type Watcher interface {
+	// Add starts watching a file for changes.
+	Add(file string) error
+
+	// Remove stops watching a file for changes.
+	Remove(file string) error
+
+	// Next blocks until a file system event occurs or the context is canceled.
+	// It returns the next event or an error if the watcher encountered a problem.
+	Next(context.Context) (*fsnotify.Event, error)
+
+	// Close shuts down the watcher and releases any resources.
+	Close() error
+}
+
+type fsnotifyWatcher struct {
+	*fsnotify.Watcher
+	closed chan struct{}
+}
+
+func NewFSNotify() (Watcher, error) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, xerrors.Errorf("create fsnotify watcher: %w", err)
+	}
+	return &fsnotifyWatcher{
+		Watcher: w,
+		closed:  make(chan struct{}),
+	}, nil
+}
+
+func (f *fsnotifyWatcher) Add(file string) error {
+	if err := f.Watcher.Add(file); err != nil {
+		return xerrors.Errorf("add file to watcher: %w", err)
+	}
+	return nil
+}
```

Comment:
https://pkg.go.dev/github.com/fsnotify/fsnotify#hdr-Watching_files-Watcher_Add

> Watching individual files (rather than directories) is generally not recommended as many programs (especially editors) update files atomically: it will write to a temporary file which is then moved to destination, overwriting the original (or some variant thereof). The watcher on the original file is now lost, as that no longer exists.

> The upshot of this is that a power failure or crash won't leave a half-written file.

> Watch the parent directory and use Event.Name to filter out files you're not interested in. There is an example of this in cmd/fsnotify/file.go.